diff --git a/openlayers.module b/openlayers.module
index 041d4a8..7979ff3 100644
--- a/openlayers.module
+++ b/openlayers.module
@@ -455,3 +455,12 @@ function openlayers_i18n_string_remove($name, $string = NULL, $options = array()
     return i18n_string_remove($name, $string, $options);
   }
 }
+
+/**
+ * Implements hook_stream_wrappers_alter().
+ */
+function openlayers_stream_wrappers_alter(&$wrappers) {
+  // Initalize the container
+  ServiceContainer::init();
+}
+
